### PR TITLE
[SPARK-37169][SQL] Fix incorrect  cast value when cast DateType to NumericType

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
In current Spark, when cast DateType to NumericType such as `cast(date'2020-01-01' as long)`, it will return `null` value, in this pr we fix this incorrect result.

### Why are the changes needed?
Fix bug

### Does this PR introduce _any_ user-facing change?
User can cast DateType to NumericType with correct result.

### How was this patch tested?
added UT

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 1]
```diff
@@ -620,8 +620,17 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
       buildCast[UTF8String](_, s => if (s.toInt(result)) result.value else null)
     case BooleanType =>
       buildCast[Boolean](_, b => if (b) 1 else 0)
+    case DateType if ansiEnabled =>
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:623]: I might be missing something but will this branch ever execute? the case DateType above would seem to cover this case too

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:623]: > I might be missing something but will this branch ever execute? the case DateType above would seem to cover this case too

All handle like timestamp type, cc @cloud-fan

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan

**Comment by Peng-Lei** [comment]
@AngersZhuuuu A little confusion, The ansi sql does not allow cast Date to numeric? I might be missing something?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:623]: This cast s not allowed in ANSI mode, why do we need this code path?

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan

**Comment by AngersZhuuuu** [comment]
> @AngersZhuuuu A little confusion, The ansi sql does not allow cast Date to numeric? I might be missing something? 

Supported now. https://github.com/apache/spark/pull/34459/files

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
also ping @gengliangwang  Since you have enable cast from date to numeric

**Pull request closed without merging**